src/s3/async_multipart_put_object.rs

```diff
@@ -0,0 +1,628 @@
+/// use cobalt_aws::config::load_from_env;
+/// use futures::AsyncWriteExt;
+///
+/// # tokio_test::block_on(async {
+/// let shared_config = load_from_env().await.unwrap();
+/// let client = Client::new(&shared_config);
+/// let dst = S3Object::new("my-bucket", "my-key");
+/// let part_size = 5 * 1_048_576;
+/// let mut writer = AsyncMultipartUpload::new(&client, &dst, part_size, None).await.unwrap();
+/// let buffer_len = 6 * 1_048_576;
+/// // Each part is uploaded as it's available
+/// writer.write_all(&vec![0; buffer_len]).await.unwrap();
+///
+/// // The pending parts are uploaded and the multipart upload is completed
+/// // on close.
+/// writer.close().await.unwrap();
+/// # })
+/// ```
+
+#[derive(Debug)]
+pub struct AsyncMultipartUpload<'a> {
+    client: &'a Client,
+    dst: S3Object,
+    upload_id: String,
+    part_size: usize,
+    max_uploading_parts: usize,
+    state: AsyncMultipartUploadState<'a>,
+}
+
+/// Minimum size of a multipart upload.
+const MIN_PART_SIZE: usize = 5_usize * MIB as usize; // 5 Mib
+/// Maximum size of a multipart upload.
+const MAX_PART_SIZE: usize = 5_usize * GIB as usize; // 5 Gib
+
+/// Default number of part which can be uploaded
+/// concurrently.
+const DEFAULT_MAX_UPLOADING_PARTS: usize = 100;
+
+impl<'a> AsyncMultipartUpload<'a> {
+    /// Create a new [AsyncMultipartUpload].
+    ///
+    /// * `client`              - S3 client to use.
+    /// * `bucket`              - Bucket to write the object into.
+    /// * `key`                 - Key to write the object into.
+    /// * `part_size`           - How large, in bytes, each part should be.
+    /// * `max_uploading_parts` - How many parts to upload concurrently.
+    #[instrument(skip(client))]
+    pub async fn new(
+        client: &'a Client,
+        dst: &S3Object,
+        part_size: usize,
+        max_uploading_parts: Option<usize>,
+    ) -> anyhow::Result<AsyncMultipartUpload<'a>> {
+        event!(Level::DEBUG, "New AsyncMultipartUpload");
+        if part_size < MIN_PART_SIZE {
+            anyhow::bail!("part_size was {part_size}, can not be less than {MIN_PART_SIZE}")
+        }
+        if part_size > MAX_PART_SIZE {
+            anyhow::bail!("part_size was {part_size}, can not be more than {MAX_PART_SIZE}")
+        }
+
+        if max_uploading_parts.unwrap_or(DEFAULT_MAX_UPLOADING_PARTS) == 0 {
+            anyhow::bail!("Max uploading parts must not be 0")
+        }
+
+        let result = client
+            .create_multipart_upload()
+            .bucket(&dst.bucket)
+            .key(&dst.key)
+            .acl(ObjectCannedAcl::BucketOwnerFullControl)
+            .send()
+            .await?;
+
+        let upload_id = result.upload_id().context("Expected Upload Id")?;
+
+        Ok(AsyncMultipartUpload {
+            client,
+            dst: dst.clone(),
+            upload_id: upload_id.into(),
+            part_size,
+            max_uploading_parts: max_uploading_parts.unwrap_or(DEFAULT_MAX_UPLOADING_PARTS),
```

Comment:
Can we update line 129 to assign this value to a variable to avoid needing to `.unwrap_or()` twice?

Reply:
I've made the validation on line 129 clearer, it should validate that 0 has not been passed in as the `max_uploading_parts`.  This remove a `.unwrap_or()`